### PR TITLE
jwt stamping

### PR DESCRIPTION
**Pull request by scottf**
https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-14.md
https://docs.nats.io/developing-with-nats/tutorials/jwt#stamping-jwt-in-languages-other-than-go

### CODE REVIEW

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (matthiashanel) [inline src/main/java/io/nats/client/support/JwtUtils.java:65]: This is missing expiration and tags. see:
https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-14.md

**Review by matthiashanel — CHANGES_REQUESTED** [review]
some of the fields are missing

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (brimworks) [inline src/main/java/io/nats/client/support/JwtUtils.java:59]: This needs to be `iat + unixSeconds`... since the duration represents how long the token should be valid relative to the issued at time.

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (brimworks) [reply src/main/java/io/nats/client/support/JwtUtils.java:89]: Agreed. It would be more clear if this took the `NKey accountSigningKey` and `NKey userKey`, and simply avoid the `String accountSeed` variant. In fact, having sensitive credentials as `String` type is not recommended:

https://stackoverflow.com/questions/8881291/why-is-char-preferred-over-string-for-passwords

If you feel this is really necessary, passing as `char[]` is preferred (although, it is trivial for users to call `NKey.fromSeed(char[])` API themselves, so I don't think it is buying us anything, and only adds to the API surface area to maintain.

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (brimworks) [inline src/main/java/io/nats/client/support/JwtUtils.java:91]: Since this is including the `userSeed`, we should be returning `char[]` rather than `String`, since this is a security best practice. Note that we should avoid any intermediate `String`s in that process. A simple way to do this is to allocate a string of the same length as the userSeed, do the `String.format()` and then convert this to a `char[]`, and then overwrite the char array offsets with the actual user seed.

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (brimworks) [inline src/main/java/io/nats/client/support/JwtUtils.java:38]: Note that all of these `JsonUtils` are not properly handling JSON string encoding for characters that need to be escaped. For example `"`, `\` or any control character such as `\n`. I would recommend asserting that all `String` inputs to APIs validate that these special characters are not found... or properly encode strings. Without this, a tag with a Windows fileshare (for example) would cause an invalid JWT to be minted and cause much confusion since no error would occur until the JWT was actually used.

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (brimworks) [inline src/main/java/io/nats/client/support/JwtUtils.java:138]: I'd suggest leaving these as private and use the more standard `toBase64Url` (lower-case first) convention.

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (brimworks) [inline src/main/java/io/nats/client/support/JwtUtils.java:132]: I have similar JWT issuing code, but the signature is computed off the `encBody` only and does NOT include the ENCODED_CLAIM_HEADER or dot. Did something change about the spec recently?

More specifically, I thought this is what the code needed to be:

```java
String encSig = ToBase64Url(accountSigningKey.sign(encBody));
```

...but I also use the `ed25519` "alg" in the JWT header, and NOT `ed25519-nkey`. So, perhaps the spec changed between these two algorithms?

**Review by brimworks — CHANGES_REQUESTED** [review]
I think there are a few things that need to be addressed before this can be merged:
* `exp` field needs to be `iat` time plus the duration when JSON serializing.
* We need to validate that an invalid JSON is not generated for String inputs.
* We should be using `char[]` and NOT `String` when dealing with sensitive credentials.

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (brimworks) [reply src/main/java/io/nats/client/support/JwtUtils.java:132]: I see that https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-14.md explicitly mentions this new way of computing the signature, so I'm assuming this is just the design of the new `ed25519-nkey` algorithm.

**Comment by scottf** [comment]
> I think there are a few things that need to be addressed before this can be merged:
> 
> * `exp` field needs to be `iat` time plus the duration when JSON serializing.
> * We need to validate that an invalid JSON is not generated for String inputs.
> * We should be using `char[]` and NOT `String` when dealing with sensitive credentials.

- `exp` yes needs to be fixed
- invalid json, the json util is dumb. This should be fixed in the util
- char, I've seen this before and I don't really get why it's more secure. In fact I think I debunked this once already, but I could be convinced

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (scottf) [reply src/main/java/io/nats/client/support/JwtUtils.java:38]: yeah the json utils assumed data from the server or from the client, so this type of checking was unnecessary. Not sure how to address this, but it needs addressing

File: src/main/java/io/nats/client/NKey.java
Comment (scottf) [reply src/main/java/io/nats/client/NKey.java:228]: I'm okay with moving this to a support class

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (scottf) [reply src/main/java/io/nats/client/support/JwtUtils.java:104]: I think b/c of it's never been used in the client so far, that we  should avoid using optional. I think I would have been better served by java doc and not naming them that. If you like, I have an entire monologue on demonstrating why "optional is better than null" is bs.

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (scottf) [reply src/main/java/io/nats/client/support/JwtUtils.java:75]: will fix

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (scottf) [reply src/main/java/io/nats/client/support/JwtUtils.java:103]: We don't need no stinkin' documentation! Uhh.... I mean yes, will do.

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (scottf) [reply src/main/java/io/nats/client/support/JwtUtils.java:59]: @brimworks can you sort this out with @matthiashanel

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (scottf) [reply src/main/java/io/nats/client/support/JwtUtils.java:138]: I've been in .NET too much lately, yes please fix. Also maybe belongs in the utility with the base32 function.

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (scottf) [reply src/main/java/io/nats/client/support/JwtUtils.java:132]: defer to Matthias

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (matthiashanel) [reply src/main/java/io/nats/client/support/JwtUtils.java:59]: agreed on `iat + unixSeconds`
from the adr:
 // expiration (when specified) time in unix seconds, derived from provided duration.

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (matthiashanel) [reply src/main/java/io/nats/client/support/JwtUtils.java:91]: My java is rusty, but I'm buying the wiping argument. (can't manually overwrite a string to erase the content)

As far as I see it, either move to []byte or get rid of the `issueUserCreds` calls, making sure to make `NATS_USER_JWT_FORMAT` public 
@brimworks what do you think about dripping `issueUserCreds` and providing the printf pattern instead?

perhaps call out the printf pattern in the doc to `issueUserJWT`

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (matthiashanel) [reply src/main/java/io/nats/client/support/JwtUtils.java:38]: Does java have no native object to json serializer?

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (matthiashanel) [inline src/main/java/io/nats/client/support/JwtUtils.java:107]: Should these raise an exception of sorts? 
* signingKey, is a mandatory account nkey pair to sign the generated jwt.
 * accountId, is a mandatory public account nkey. Will return error when not set or not account nkey.
 * publicUserKey, is a mandatory public user nkey. Will return error when not set or not user nkey.

**Review by matthiashanel — CHANGES_REQUESTED** [review]

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (scottf) [reply src/main/java/io/nats/client/support/JwtUtils.java:38]: java does not and we've been asked to avoid 3rd party libraries

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (brimworks) [reply src/main/java/io/nats/client/support/JwtUtils.java:91]: Using char[] rather thing String is a security best practice. The problem with String is that the garbage collector works differently in regards to Strings, and strings may even be "intern"ed so they get reused. Anyways, if you all feel strongly about using String, that is fine, I'm just bringing this up as a Java best-practice.

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (brimworks) [reply src/main/java/io/nats/client/support/JwtUtils.java:104]: >  I have an entire monologue on demonstrating why "optional is better than null" is bs.

I'd love to hear it. I personally find Optional to be a great tool. Especially if used in combination with static code analysis.

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (brimworks) [reply src/main/java/io/nats/client/support/JwtUtils.java:59]: Sounds like we do need the `iat + unixSeconds`.

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (brimworks) [reply src/main/java/io/nats/client/support/JwtUtils.java:38]: Based on conversation off-line, we can easily fix this by modifying `JsonUtils` so these characters are properly encoded.

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (matthiashanel) [reply src/main/java/io/nats/client/support/JwtUtils.java:91]: @brimworks What I mean is basically avoid the entire problem by not having `issueUserCreds` and in the documentation of `issueUserJWT` point at `NATS_USER_JWT_FORMAT` instead.

Comment (brimworks) [reply src/main/java/io/nats/client/support/JwtUtils.java:91]: Yeah. I do like that idea. @scottf do you have an opinion?

Comment (scottf) [reply src/main/java/io/nats/client/support/JwtUtils.java:91]: I'm fine with whatever you guys want. If It satisfies Matthias's requirements then go for it. We can always add it in later, it's harder to take it back

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (brimworks) [reply src/main/java/io/nats/client/support/JwtUtils.java]: I just removed this method to avoid the char[] vs String issue.

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (brimworks) [reply src/main/java/io/nats/client/support/JwtUtils.java]: I've changed this to public.

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (brimworks) [reply src/main/java/io/nats/client/support/JwtUtils.java]: done

File: src/main/java/io/nats/client/NKey.java
Comment (brimworks) [reply src/main/java/io/nats/client/NKey.java]: Moved to RandomUtils support class... since it seems to contain an assortment of utilities such as bytesToLong() and nextLong(Random, long). I suspect that class was originally intended for actual random number generation type of stuff, but bytesToLong() has nothing to do with random generation... so it seems reasonable.

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (brimworks) [reply src/main/java/io/nats/client/support/JwtUtils.java]: `issueUserCreds` was removed and the `NATS_USER_JWT_FORMAT` is public along with documentation on how to use it to create credentials.

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (brimworks) [reply src/main/java/io/nats/client/support/JwtUtils.java]: `JsonUtils` now properly escapes string characters that need escaping per the spec.

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (brimworks) [reply src/main/java/io/nats/client/support/JwtUtils.java]: Update to the more standard camelCase.

File: src/main/java/io/nats/client/support/JwtUtils.java
Comment (brimworks) [reply src/main/java/io/nats/client/support/JwtUtils.java:158]: Yes. All of these are now validated and have unit tests for the validation.

**Review by brimworks — APPROVED** [review]
I've fixed on the comments I had in regards to this PR, so marking it as approved.

File: src/test/java/io/nats/client/support/JwtUtilsTests.java
Comment (matthiashanel) [inline src/test/java/io/nats/client/support/JwtUtilsTests.java:38]: just to be sure, the `\n` at the end is an artefact of the test, not the jwt.
I stored this string without the `\n` and nsc is happy. 
```
nsc describe jwt -f x
+---------------------------------------------------------------------------+
|                                   User                                    |
+----------------+----------------------------------------------------------+
| Name           | UA6KOMQ67XOE3FHE37W4OXADVXVYISBNLTBUT2LSY5VFKAIJ7CRDR2RZ |
| User ID        | UA6KOMQ67XOE3FHE37W4OXADVXVYISBNLTBUT2LSY5VFKAIJ7CRDR2RZ |
| Issuer ID      | ADQ4BYM5KICR5OXDSP3S3WVJ5CYEORGQKT72SVRF2ZDVA7LTFKMCIPGY |
| Issuer Account | ACXZRALIL22WRETDRXYKOYDB7XC3E7MBSVUSUMFACO6OM5VPRNFMOOO6 |
| Issued         | 2021-09-30 23:09:38 UTC                                  |
| Expires        |                                                          |
| Issuer Scoped  | Yes                                                      |
+----------------+----------------------------------------------------------+
```

**Review by matthiashanel — COMMENTED** [review]
I think this looks good. I'll leave the final say on this with @scottf

**Review by matthiashanel — APPROVED** [review]
LGTM

**Pull request merged**
